### ISSUE TITLE
Implement accessibility checking in CI

### ISSUE THREAD

**Issue by brainwane** [SolutionGuidance/psm#518]
Integrate automated checking of errors in WCAG Level A and Level AA standards in our continuous integration. This will probably require wiring something into Serenity/Selenium so that the tool can simulate logging in, filling in forms, and navigating; automated accessibility (a11y) tools often look for elements within the rendered HTML and the CSS/JS. [axe-core](https://github.com/dequelabs/axe-core) and [pa11y-ci](https://github.com/pa11y/pa11y-ci) seem promising.

Context: as discussed in #415, we [need to ensure](https://www.access-board.gov/guidelines-and-standards/communications-and-it/about-the-ict-refresh/overview-of-the-final-rule) that the PSM meets Level A and Level AA Success Criteria and Conformance Requirements in WCAG 2.0 in order to meet the requirements of [Section 508 of the Rehabilitation Act of 1973](https://www.section508.gov/content/learn/laws-and-policies) and, more broadly, [the Department of Justice's interpretation of the Americans with Disabilities Act Title II Regulations](https://www.ada.gov/regs2010/titleII_2010/titleII_2010_regulations.htm).

**Comment by brainwane** [comment]
I'm going to leave a few notes here that I found interesting to at least skim when I was researching this:

* http://cruft.io/posts/automated-accessibility-testing-node-travis-ci-pa11y/
* https://bitsofco.de/pa11y/
* http://hollsk.co.uk/posts/view/using-actions-in-pa11y
* https://github.com/dequelabs/axe-selenium-java
* https://marcysutton.github.io/a11y-automated-testing/
* https://www.deque.com/blog/accessibility-testing-axe-webdriverjs/
* https://saucelabs.com/blog/accessibility-ax-testing-in-the-devops-chain

(I couldn't test [`axe-core`](https://github.com/dequelabs/axe-core#getting-started) easily when I was investigating our options, because it wants you to "Now include the javascript file in each of your iframes in your fixtures or test systems" and "Now insert calls at each point in your tests where a new piece of UI becomes visible or exposed". But axe might be the right choice for us.)

**Comment by PaulMorris** [comment]
Here is one approach.  Basically, set up Selenium tests to run the [HTML_CodeSniffer](https://squizlabs.github.io/HTML_CodeSniffer/) bookmarklet code on each page of the site and check the results that it gives.

The HTML_CodeSniffer bookmarklet injects some JS into the current page, and that JS runs [this remote HTMLCS.js script](https://squizlabs.github.io/HTML_CodeSniffer/build/HTMLCS.js) on the page.  Similarly, we can use `pageObject.evaluateJavascript` in Selenium tests to inject and run the HTML_CodeSniffer JS code on each page (e.g. as a Java string).  Then it is just a matter of using Selenium to read the number of accessibility errors (and possibly also what kinds of errors) from the widget that is displayed on the page.

Also, before checking the results, the Selenium tests could select the type of accessibility standard we are using from the drop down menu in the widget.

(If relying on a remote script was a concern, we could save the remote HTMLCS.js script in a local file and use that instead.)

I have worked up a rough but basically successful proof-of-concept of this approach in this [accessibility-testing](https://github.com/solutionguidance/psm/tree/accessibility-testing) branch.

**Comment by PaulMorris** [comment]
[axe-core](https://github.com/dequelabs/axe-core) is looking promising since it seems focused on integrating with automated testing tools like Selenium.  So it should work well for integrating with our existing Selenium integration test setup, allowing us to leverage our existing test code for logging in, navigating, and interacting with the site.  For example:

https://axe-core.org/examples/selenium.html
https://www.deque.com/blog/accessibility-testing-axe-webdriverjs/
https://github.com/dequelabs/axe-webdriverjs
https://github.com/dequelabs/axe-selenium-java
https://axe-core.org/examples/

There are axe-core browser add-ons that allow for manual checking, that integrate into browser dev tools, and the API allows for filtering by various accessibility standards, including the ones we are targeting: https://axe-core.org/docs/#api-name-axegetrules

**Comment by PaulMorris** [comment]
More on [axe-core](https://github.com/dequelabs/axe-core) and some questions.

Eventually we'd use npm to manage the axe-core dependency (as part of #626).
For now, do we just use the axe.js file as with our other JS dependencies, until npm is set up?

The aXe APIs are provided in the javascript file axe.js.  "It must be included in the web page under test. Parameters are sent as javascript function parameters.  Results are returned in JSON format."

Two possibilities for making the axe.js file accessible on each page.

1. Either inject the contents of the axe.js file into each page as part of the tests, using the JS injection feature.  (How to access axe.js and read it into a Java string?)

2. Set up some kind of build flag for testing that adds a `<script>` tag to the top-level templates, to include axe.js on each page only when testing/debugging.

We then call axe.run(context, options, callback) on the page.  
"options" can be used to specify the standard we're checking against, etc.  
"callback" is a function that receives the results.
If callback is not defined, a promise is returned (might be better!?).

The axe.run() call could be made by injecting the JS call into the page and assigning the results to a variable that the test can then do an assertion against.  Most basically we test to see if the violations count is 0.  There should be a way to report the details of the failure when a test fails, as these are supplied in the JSON that axe.run() returns.

How to organize the selenium tests?  
@jasonaowen  and I talked about this briefly a while ago.  We'd like to put accessibility tests into their own grouping of tests, to be run separately.  May entail some rearrangement of the testing setup.

We'll write the tests using Cucumber like our other Selenium tests.  Looking at their Ruby plus Cucumber integration might be helpful:
https://github.com/dequelabs/axe-matchers
https://github.com/dequelabs/axe-matchers/blob/master/docs/Cucumber.md

Errors vs Warnings distinction as in HTML_CodeSniffer?  
I haven't seen this distinction yet with axe-core.  So far we've focused on fixing errors not warnings, so this will give us more violations to fix.

**Comment by PaulMorris** [comment]
The axe.run() call is asynchronous (either calling a supplied callback function or returning a promise) and returns JS/JSON data that we'll need to assert against.  These are two things we haven't handled yet in our Selenium tests.  

Selenium has [executeAsyncScript](https://seleniumhq.github.io/selenium/docs/api/java/org/openqa/selenium/JavascriptExecutor.html) that takes care of both of these things. A callback is called when the async call is done, and the results are returned from that callback and can be captured in a Java JSONObject or JSONArray.  Given that axe.run() takes its own callback function, it's not clear how the two callbacks relate.  So I looked into [axe-selenium-java](https://github.com/dequelabs/axe-selenium-java) to see how it is used there.

That helped, but made me think we may want to use axe-selenium-java.  It is targeted at JUnit rather than Cucumber, but that should be no problem, since the AXE helper class and methods it provides can be used with either type of tests.  It uses Maven (rather than Gradle) in its docs. Also it is injecting the axe-core code into the page, rather than putting it in a `<script>` tag, which requires making that file accessible through a Maven dependency setup (a POM file which seems to be a Maven thing).

So we could either use this helper library (assuming we can use it via Gradle instead of Maven), or just borrow some of its approach as we write our own code to work with Serenity directly.

**Comment by PaulMorris** [comment]
After conferring with @jasonaowen about [axe-selenium-java](https://github.com/dequelabs/axe-selenium-java) we are proceeding with it, see WIP PR #639.

**Comment by PaulMorris** [comment]
Here are the pages that will still not have a11y tests after PRs #676, #679, #684, and #686 land.

**For service admin login**
Pages for individual 'view enrollment' 
Pages for organization and individual 'edit enrollment'
Pages for organization and individual 'renew enrollment'
My Profile page
Update profile page
Advanced Search page
Quick Search results page

**For system admin login**
Providers page
Service Agents page
Service Administrators page
System Administrators page
View User Account page
Edit User Account Page
Delete User Account Modal 
Quick Search Results
Advanced Search Results

**Comment by cecilia-donnelly** [comment]
Accessibility checking has been added to our continuous integration setup!  Thank you, @PaulMorris.  I've opened #709 to track future a11y tests.
